### ISSUE TITLE
Optimize request queue and detect closed connections from client in netty server

### ISSUE THREAD

**Issue by tejas2hit** [reactor/reactor-netty#2658]
We are using netty server in Sprint Boot. 

When load is high on our server we sometimes see error with message: `context canceled`. 
Our assumption is that this happens as there is timeout set on client side, when this timeout occurs the request is cancelled from client side. In these scenarios no response is sent from our side, however we are still processing the request and finishing.

We would like detect such cancelled request, stop processing them and send a default response with 500.

Is there any configuration in netty where we could detect such cancelled request and stop processing them?

# Environment:
reactor-netty version: 4.1.86
spring-boot: 2.7.7
spring-boot-starter-webflux: 2.7.7
JDK: 17

**Comment by pderop** [comment]
@tejas2hit ,

Thanks for getting in touch, it feels like this is a question that would be better suited to [Gitter](https://gitter.im/reactor/reactor-netty) or [Stack Overflow](https://stackoverflow.com/questions/tagged/reactor-netty). As mentioned in [README](https://github.com/reactor/reactor-netty/#getting-help), we prefer to use GitHub issues only for bugs and enhancements.

Now, for this particular question, I will need more informations:

-  are you using a **netty** server, or a **reactor-netty** server ?
- you indicated _reactor-netty version 4.1.86_, but this is the version for netty, so what is the version for reactor-netty (if you are using it) ?
- what protocol is used ? HTTP 1.1, or H2 ? or H2C ?
- when you see "context canceled", where do you see the log ? from the server or from the client ?
- can you elaborate more about the use case ? is the following correct:  You have one client jvm which is sending a request to the server (and the server is using reactor-netty ? or netty ?). The client uses the webflux WebClient, and because the load is high, the WebClient times out. But the server is still handling the request, and you need to get the server notified quickly that the request is aborted, in order to send back a 500 error to the client. But then, if the client has closed the connection, how the server can send back the 500 error ?
- Or, maybe it's a more complex scenario, where the server receives a client request, forwards it to another upstream server using the webflux WebClient, which times out, but then the server is not notified about the cancelled webclient request and you would like the server to immediately respond 500 to the origin client in case the forwarded request times out ?

so, can you clarify ? maybe provide a minimal reproducer sample project which we can build and test ? it would help to understand the scenario ?

thanks.

**Comment by piyushmor** [comment]
we are using `io.projectreactor.netty:reactor-netty-http:1.0.26` the problem is that we see excessive queueing up of requests on our service even after client has timed out. we would like to ensure that we return a 5xx after some specified timelimit to ensure we don't have a cascading effect due to this queueing up of requests.

**Comment by pderop** [comment]
There is a micrometer meter in reactor netty which reports the current Netty eventloop task queue size, so that you can come to know if the queue is starting to pill up and the server is then entering into a state where requests can't be responded timely.

Is it this kind of information you need ? 

The metric name is `reactor.netty.eventloop.pending.tasks`(check https://projectreactor.io/docs/netty/release/reference/index.html#getting), and reports the number of tasks that are pending for processing on an event loop. This metric actually corresponds to the pendingTask method from the netty [SingleThreadEventExecutor.pendingTask](https://netty.io/4.1/api/io/netty/util/concurrent/SingleThreadEventExecutor.html#pendingTasks--
) method.

Here is a simple http server using reactor-netty (micrometer is not used, because it's easy and efficient to get access to the current request event loop and then simply obtain the pending task.
The requests are rejected using a simple progressive overload controller: when the current eventloop pending tasks are starting to cross a given low water mark, then requests are gradually rejected, and when the pending tasks are crossing a given high water mark, then all requests are rejected.

Is it this kind of thing you would need ? then why not implementing your own overload controller based on the pending task information ?

```java
import reactor.netty.http.HttpProtocol;
import reactor.netty.http.server.HttpServer;
import reactor.netty.http.server.HttpServerRequest;

import java.time.Duration;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.atomic.AtomicReference;

public class FlowControlServer {

    /**
     * Progressive overload controller: request are gradually rejected when we the number of current event 
     * loop pending tasks are crossing a given low waterk. Then when the high water mark is crossed, all 
     * requests are rejected. This simple mechanism allows to progressively reject requests when the event 
     * loops are becoming very busy.
     */
    public static class FlowControl {
        protected final int lowWM, highWM, interval;

        public FlowControl(int lowWM, int highWM) {
            this.lowWM = lowWM;
            this.highWM = highWM;
            interval = this.highWM - this.lowWM;
        }

        public boolean isOverloaded(int pendingTasks) {
            if (pendingTasks > lowWM) {
                if (pendingTasks > highWM) {
                    return false;
                }
                int overload = ((pendingTasks - lowWM) * 100) / interval;
                return ThreadLocalRandom.current().nextInt(100) >= overload;
            }
            return true;
        }
    }

    public static void main(String[] args) {
        FlowControlServer server = new FlowControlServer();
        server.start();
    }

    private void start() {
        try {
            System.out.println("Starting server@");
            FlowControl flowControl = new FlowControl(100, 200);

            DisposableServer server =
                    HttpServer.create()
                            .option(ChannelOption.SO_REUSEADDR, true)
                            .protocol(HttpProtocol.HTTP11, HttpProtocol.H2C)
                            .port(8080)
                            .handle((req, res) -> {
                                if (flowControl.isOverloaded(getPendingTask(req))) {
                                    return res.status(500).send();
                                }
                                return req.receive()
                                        .aggregate()
                                        .asString()
                                        .log("data")
                                        .then(res.status(200)
                                                .sendString(Mono.just("OK")
                                                        .delayElement(Duration.ofMillis(10)))
                                                .neverComplete());
                            })
                            .wiretap(false)
                            .bindNow();

            server.onDispose().block();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private int getPendingTask(HttpServerRequest req) {
        AtomicReference<SingleThreadEventExecutor> ref = new AtomicReference<>();
        req.withConnection(conn -> {
            EventLoop eventLoop = conn.channel().eventLoop();
            if (eventLoop instanceof SingleThreadEventExecutor) {
                ref.set((SingleThreadEventExecutor) eventLoop);
            }
        });
        SingleThreadEventExecutor eventLoop = ref.get();
        return eventLoop == null ? 0 : eventLoop.pendingTasks();
    }
}
```

using the above, my h2load loader reported 2225290 succesful requests, and 14893 failed requests (the failures corresponds to the state where the eventloop pending tasks are starting to cross [100, 200];

**Comment by piyushmor** [comment]
instead of having this based on count, can we do this based on the amount of time for which the request is staying queued? Also, is there a lifo mechanism available for the incoming requests?

**Comment by pderop** [comment]
@piyushmor ;

Full monitoring of the delay execution for scheduled tasks may cause excessive performance loss, but there is this old Netty issue (still opened) which sounds promising and it seems to provide the kind of metric you are looking for (in an efficient way):

https://github.com/netty/netty/pull/9080

Give me some time, I need to check this Netty issue and I'll get back to you ...
thanks

**Comment by piyushmor** [comment]
can the code you shared above be adapted to work on this metric to start rejecting requests?

**Comment by pderop** [comment]
yes, this is the intent, but before, I need to check the state of the old Netty PR, will start to work on this today.

**Comment by violetagg** [comment]
@piyushmor Please clarify whether you are interested in the tasks duration in the event loop (task in the event loop can be everything) or your are interested in the pending requests. Because if it is related to the requests, the solution might be different.

**Comment by pderop** [comment]
@tejas2hit,

Before going ahead, I need some clarifications about the following, in order to avoid going into a bad direction.

I have two questions:

1) When you say:

> Our assumption is that this happens as there is timeout set on client side, when this timeout occurs the request is cancelled from client side. In these scenarios no response is sent from our side, however we are still processing the request and finishing.
> We would like detect such cancelled request, stop processing them and send a default response with 500.

So, when there is a timeout on the client, the connection will be closed, then it won't be possible to send a 500 error. Can you clarify ?

2) when saying:

> When this timeout occurs the request is cancelled from client side. In these scenarios no response is sent from our side, **however we are still processing the request and finishing.**

Then if the handling of a request is doing a long processing, then do you try to interrupt the processing, for example, by hooking on the connection close when you receive the request; so I mean that maybe you see many processed requests because they are not cancelled at the time the client is closing the connection ? Can you check that please ?

**Comment by tejas2hit** [comment]
@pderop 

1. For this:

> So, when there is a timeout on the client, the connection will be closed, then it won't be possible to send a 500 error. Can you clarify ?

Yes, that is correct. I realised after first reply that it won't be possible to send response as connection is closed from client side.

2. For:

> Then if the handling of a request is doing a long processing, then do you try to interrupt the processing, for example, by hooking on the connection close when you receive the request; so I mean that maybe you see many processed requests because they are not cancelled at the time the client is closing the connection ? Can you check that please ?

We do have timeouts configured for some of the processes, but not for all.  We do see errors:
Error [java.lang.UnsupportedOperationException] for HTTP GET "/our/api", but ServerHttpResponse already committed (200 OK)
Which might be related but not sure what to make of this error.

**Comment by pderop** [comment]
> We do have timeouts configured for some of the processes, but not for all. We do see errors:
> Error [java.lang.UnsupportedOperationException] for HTTP GET "/our/api", but ServerHttpResponse already committed (200 OK)
> Which might be related but not sure what to make of this error.

Sorry to insist, but I'm not sure I follow. Let me rephrase my second question: let's say you have this kind of server:

```java
    disposableServer = createServer(9999)
	.handle((req, resp) -> resp.sendString(Mono.<String>create(sink -> {
		AtomicBoolean cancelled = new AtomicBoolean(false);
		resp.withConnection(conn -> conn.onDispose(() -> { // (1)
		    cancelled.set(true);
		    sink.error(new Exception("operation interrupted"));
		}));
		for (int i = 0; i < 1000 && !cancelled.get(); i++) { // (2)
		    // Do a cpu intensive task
		}
		sink.success("response");
	 })
	.subscribeOn(Schedulers.boundedElastic()) // (3)
	).neverComplete()
    ).bindNow();
```

Here, we have a server which schedules the handling of requests using the `Schedulers.boundedElastic`, see `(3)` because the handling of the request is CPU intensive, and you don't want to overload the netty event loop thread.

Now, if the client closes the connection, the above examples uses a "cancelled" flag which is set to true when the connection is disposed, see `(1)`, and the for/loop then quickly breaks if the cancelled flag is set to true, see `(2)`.

So,  I'd like to know if you are in such kind of scenario, and if so, then do you cancel the processing of requests, else this would explain why you observe many requests still being processed  after the client has closed the connections ...

**Comment by tejas2hit** [comment]
@pderop Thank you for your reply. This seems close to our scenario. We have more network intensive tasks than CPU intensive. 
However, we do not cancel the request processing when connection it cancelled from client, so that probably is the reason for requests still being processed.  
When I use `.handle((req, resp)` the response from Spring Boot controller is neglected. 
Is there a way around to use this solution in Spring Boot?

**Comment by pderop** [comment]
@tejas2hit ,

Please provide a kind of sample code which shows what is doing the server from the `handle((req, resp))` callback ?
I assume that the server `handle((req, resp))` method sends an http request to another upstream server, maybe using WebCLient from webflux, but I even don't know ...

Can you provide such sample code please ?

**Comment by piyushmor** [comment]
> @tejas2hit,
> 
> Before going ahead, I need some clarifications about the following, in order to avoid going into a bad direction.
> 
> I have two questions:
> 
> 1. When you say:
> 
> > Our assumption is that this happens as there is timeout set on client side, when this timeout occurs the request is cancelled from client side. In these scenarios no response is sent from our side, however we are still processing the request and finishing.
> > We would like detect such cancelled request, stop processing them and send a default response with 500.
> 
> So, when there is a timeout on the client, the connection will be closed, then it won't be possible to send a 500 error. Can you clarify ?
> 
> 2. when saying:
> 
> > When this timeout occurs the request is cancelled from client side. In these scenarios no response is sent from our side, **however we are still processing the request and finishing.**
> 
> Then if the handling of a request is doing a long processing, then do you try to interrupt the processing, for example, by hooking on the connection close when you receive the request; so I mean that maybe you see many processed requests because they are not cancelled at the time the client is closing the connection ? Can you check that please ?

It seems like a disconnect between netty and spring boot. Spring tries to send a response code after processing but the connection is already closed. What we would like to achieve is that the task is cancelled completely if the connection is already closed by the peer. Do you have any pointers for that? we are using Spring Boot Webflux with kotlin coroutines, so not sure if that has some impact on this functionality.

**Comment by piyushmor** [comment]
> @piyushmor Please clarify whether you are interested in the tasks duration in the event loop (task in the event loop can be everything) or your are interested in the pending requests. Because if it is related to the requests, the solution might be different. And please specify what you use - HTTP/1.1, HTTP/1.1 pipelining, HTTP/2 as it is important.

we are using HTTP/1.1  and we want to only check how long incoming requests are waiting before Spring starts working on them.

**Comment by pderop** [comment]
@piyushmor ,

> It seems like a disconnect between netty and spring boot. Spring tries to send a response code after processing but the connection is already closed. What we would like to achieve is that the task is cancelled completely if the connection is already closed by the peer. Do you have any pointers for that? we are using Spring Boot Webflux with kotlin coroutines, so not sure if that has some impact on this functionality.

Again, please provide a minimal sample code so we continue to discuss and go ahead.
I would like for example the following kind of sample code where we have a reactor-netty server which uses a webclient from the handle method (**but I don't know if it's your use case** ?)

In this example, if the origin client closes the connection, then the webclient will be also cancelled, and from the `doOnCancel()` callback you can also possibly perform any needed cleanup actions:

```java
        HttpClient client = HttpClient.create()
                .responseTimeout(Duration.ofMillis(1000))
                .wiretap(true);

        DisposableServer server =
                HttpServer.create()
                        .option(ChannelOption.SO_REUSEADDR, true)
                        .protocol(HttpProtocol.HTTP11)
                        .port(8080)
                        .handle((req, res) -> res.sendString(webClient
                                .get()
                                .retrieve()
                                .bodyToMono(String.class)
                                .doOnCancel(() -> log.warn("WebClient has been cancelled"))
                                .onErrorResume(WebClientRequestException.class, exception -> {
                                    if (exception.getCause() instanceof ReadTimeoutException) {
                                        return Mono.just("request timeout");
                                    } else {
                                        return Mono.just("Internal error: " + exception);
                                    }
                                })))
                        .bindNow();

        server.onDispose().block();
```

**Comment by tejas2hit** [comment]
@pderop Thank you for the response. It gave some idea about our scenario. Here is somewhat similar scenario with sample code within Spring Boot application:

`
 lateinit var scope: CoroutineScope

    @GetMapping("/hello")
    suspend fun getCall(): String {

        scope = CoroutineScope(coroutineContext)
        val dispatcher = Dispatchers.IO + SupervisorJob()

        val r1 = scope.async(dispatcher){
            externalService.getActivity()  // API call with webclient to other service
                .doOnCancel {
                    println("R1 cancelled")
                }.doOnSuccess {
                    println("R1 success")
                }.awaitFirstOrNull()
        }

        val r2 = scope.async(dispatcher){
            externalService.getActivity()   // API call with webclient to other service
                .doOnCancel {
                    println("R2 cancelled")
                }.doOnSuccess {
                    println("R2 success")
                }.awaitFirstOrNull()
        }

        val r3 = scope.async(dispatcher){
            externalService.getActivity()   // API call with webclient to other service
                .doOnCancel {
                    println("R3 cancelled")
                }.doOnSuccess {
                    println("R3 success")
                }.awaitFirstOrNull()
        }

        return r1.await().plus(r2.await()).plus(r3.await())
    }
`

When I make a request which would timeout to `/hello`, I get following response.
`R1 success
R2 success
R3 success
R2 cancelled
R1 cancelled
R3 cancelled`

All requests have succeeded even though it was cancelled from client side.

When I experimented more I observed if I remove `+ SupervisorJob()` from this code then request is correctly cancelled and I get following response:
`R1 cancelled
R2 cancelled
R3 cancelled`

This seems to be happening because with _SupervisorJob()_ failure/cancellation of one child does not affect the other children.
Ref: https://victorbrandalise.com/coroutines-part-ii-job-supervisorjob-launch-and-async/

Also, it seems even if the request is successfully finished, it invokes `doOnCancel()` in both cases.
Overall, using `SupervisorJob()` seems to be reason that requests are not cancelled. 
Please share your thoughts.

**Comment by pderop** [comment]
> When I make a request which would timeout to /hello, I get following response.
> R1 success R2 success R3 success R2 cancelled R1 cancelled R3 cancelled
> 
> All requests have succeeded even though it was cancelled from client side.
> 

-> can you provide a minimal kotlin reproducer project which I can build and test (either using gradle or maven) ?

> When I experimented more I observed if I remove + SupervisorJob() from this code then request is correctly cancelled and I get following response:
R1 cancelled R2 cancelled R3 cancelled

-> I'm not a kotlin user, but if what you want is to cancel all jobs when one particular job is cancelled or fails, then why are you using `SupervisorJob` ? Looking at https://victorbrandalise.com/coroutines-part-ii-job-supervisorjob-launch-and-async/, it indeed seems that `SupervisorJob` does not cancel all jobs when one particular job is cancelled or has failed.
so, why do you need to use such `SupervisorJob` ? please clarify ?
 
Now, if you really need to use `SupervisorJob`, then maybe you need to cancel the scope manualy from doOnCancel callbacks ? Also, notice that if one webclient activity fails because the upstream destination server has closed the socket or because a 500 internal server error is returned by the destination server, then the `doOnError` callback will be invoked. So, if cancelling the `scope` really cancels everything and if you really need to use `SupervisorJob`, then maybe what you would need is something like this:

```kotlin
@GetMapping("/hello")
suspend fun getCall(): String {
    val scope = CoroutineScope(coroutineContext)
    val dispatcher = Dispatchers.IO + SupervisorJob()
    val r1 = scope.async(dispatcher){
        externalService.getActivity()  // API call with webclient to other service
            .doOnCancel {
                println("R1 cancelled")
                scope.cancel();
            }.doOnError {
                println("R1 failed")
                scope.cancel();
            }.doOnSuccess {
                println("R1 success")
            }.awaitFirstOrNull()
    }
    val r2 = scope.async(dispatcher){
        externalService.getActivity()   // API call with webclient to other service
            .doOnCancel {
                println("R2 cancelled")
                scope.cancel();
            }.doOnError {
                println("R2 failed")
                scope.cancel();
            }.doOnSuccess {
                println("R2 success")
            }.awaitFirstOrNull()
    }
    val r3 = scope.async(dispatcher){
        externalService.getActivity()   // API call with webclient to other service
            .doOnCancel {
                println("R3 cancelled")
                scope.cancel();
            }.doOnError {
                println("R3 failed")
                scope.cancel();
            }.doOnSuccess {
                println("R3 success")
            }.awaitFirstOrNull()
    }
    return r1.await().plus(r2.await()).plus(r3.await())
}

```

**Comment by pderop** [comment]
if the client of the service calling the endpoint times out (and closes the connection), then the webclient should be called in the `doOnCancel()` callback. 

please provide a minimal sample code (one that we can build using gradle or maven) that is doing the kind of code you showed previously from https://github.com/reactor/reactor-netty/issues/2658#issuecomment-1399953380.

thanks.

**Comment by pderop** [comment]
Hi @tejas2hit ,

I'm closing this one. Of course, If you need more informations and if you are able to attach a minimal sample code that we can build and test, then feel free to reopen this issue with the attached sample code.
thank you.

**Comment by piyushmor** [comment]
we are working on the coroutine changes and will get back if the issue persists
